### ISSUE TITLE
DOMParser interprets newline character after <br> tag as leading space character in following text node

### ISSUE THREAD

**Issue by DouweM** [ProseMirror/prosemirror#861]
### Issue details

HTML containing a `<br>` tag followed by a newline character shows up incorrectly (or at least, not as expected) in ProseMirror, because parsing `"<p>Before<br>\nAfter</p>"` and serializing it again results in `"<p>Before<br> After</p>"`.

The underlying cause appears to be `DOMParser`'s whitespace normalization. Unfortunately, none of the `preserveWhitespace` options produce the desired effect. The description for that option says "By default, whitespace is collapsed as per HTML's rules.", but that doesn't appear to be correct in this case.

Should we teach `DOMParser` to ignore whitespace after `<br>`?

### Steps to reproduce

https://gossamer-station.glitch.me/ or:

```js
const schema = new Schema({
  nodes: {
    doc: {
      content: "block+"
    },

    paragraph: {
      content: "inline*",
      group: "block",
      parseDOM: [{tag: "p"}],
      toDOM() { return ["p", 0] }
    },

    text: {
      group: "inline"
    },

    hard_break: {
      inline: true,
      group: "inline",
      selectable: false,
      parseDOM: [{tag: "br"}],
      toDOM() { return ["br"] }
    }
  },
  marks: {}
});

const htmlBefore = "<p>Before<br>\nAfter</p>";

const element = document.createElement('div');
element.innerHTML = htmlBefore;

const doc = DOMParser.fromSchema(schema).parse(element);
// Expected: {"type":"doc","content":[{"type":"paragraph","content":[{"type":"text","text":"Before"},{"type":"hard_break"},{"type":"text","text":"After"}]}]}
// Actual:   {"type":"doc","content":[{"type":"paragraph","content":[{"type":"text","text":"Before"},{"type":"hard_break"},{"type":"text","text":" After"}]}]}

const htmlAfter = DOMSerializer.fromSchema(schema).serializeFragment(doc.content);
// Expected: "<p>Before<br>After</p>"
// Actual:   "<p>Before<br> After</p>"
```

Expected (`htmlBefore`):

<img width="56" alt="captura de pantalla 2018-10-24 a la s 18 30 42" src="https://user-images.githubusercontent.com/159434/47467315-76653300-d7bb-11e8-9321-8ac316d8b1a6.png">

Actual (`htmlAfter`): 
<img width="58" alt="captura de pantalla 2018-10-24 a la s 18 30 49" src="https://user-images.githubusercontent.com/159434/47467317-76fdc980-d7bb-11e8-86e1-78f219388e7a.png">

### ProseMirror version

prosemirror-model 1.6.2

**Comment by DouweM** [comment]
@marijnh I'm happy to have a go at fixing this myself if you can lead me in the right direction! Would `ParseContext.addTextNode` be the most appropriate place?

**Comment by marijnh** [comment]
Yes, that looks like the right place to fix this. I'm not quite sure how it should be formulated, though. Whitespace around an image, for example, shouldn't be dropped. I guess `<br>` is treated as an exception, like it is whitespace, in HTML?

**Comment by DouweM** [comment]
@marijnh I've submitted a fix at https://github.com/ProseMirror/prosemirror-model/pull/29, but as I mentioned there, I don't love it. Your review would be appreciated :)

**Comment by adrianheine** [comment]
I had to special-case `br`s for CodeMirror as well in codemirror/codemirror.next@f08595c49f73b2b064f0a4e9f330c127f5c7089d for newline handling.
